Core/include/Acts/TrackFitting/GlobalChiSquareFitter.hpp

```diff
@@ -728,13 +737,30 @@ class Gx2Fitter {
 
     // Calculate covariance of the fitted parameters with inverse of [a]
     BoundMatrix fullCovariancePredicted = BoundMatrix::Identity();
-    if (aMatrix.topLeftCorner<reducedMatrixSize, reducedMatrixSize>()
-            .determinant() != 0) {
-      fullCovariancePredicted
-          .template topLeftCorner<reducedMatrixSize, reducedMatrixSize>() =
-          aMatrix.topLeftCorner<reducedMatrixSize, reducedMatrixSize>()
-              .inverse();
-    } else if (gx2fOptions.nUpdateMax > 0) {
+    bool detAisZero = true;
+    if (gx2fOptions.zeroField) {
+      constexpr size_t reducedMatrixSize = 4;
+      if (aMatrix.topLeftCorner<reducedMatrixSize, reducedMatrixSize>()
+              .determinant() != 0) {
+        detAisZero = false;
+        fullCovariancePredicted
+            .topLeftCorner<reducedMatrixSize, reducedMatrixSize>() =
+            aMatrix.topLeftCorner<reducedMatrixSize, reducedMatrixSize>()
+                .inverse();
+      }
+    } else {
+      constexpr size_t reducedMatrixSize = 5;
+      if (aMatrix.topLeftCorner<reducedMatrixSize, reducedMatrixSize>()
+              .determinant() != 0) {
```

Comment:
There's a function that checks if a matrix is invertible AND produces the inverse if so. I believe this is slightly more efficient than calculating the determinant, and then calculating inverse.
Also: what happens if this is not the case, i.e. we just don't include this in the covariance?

Reply:
determinant is not a good measure of invertibility. and it is unlikely to be exactly 0 if you didn't hand craft it that way. and if it is non 0 it may still not be numerically invertible

the helper would be a good choice which paul mentioned

Reply:
I couldn't find a function that only inverts when possible, but Eigen suggests to use the following workflow:
1. Create FullPivLU Matrix $A=P^{−1}LUQ^{−1}$
2. Check invertibility `a.isInvertible()`
3. Invert `a.inverse()`

https://eigen.tuxfamily.org/dox/classEigen_1_1FullPivLU.html#a0ee7753645eb31bcbd5faa459168b294

For matrices up to 4x4 the normal/direct inverse is a bit more efficient, but for the invertibility check one needs to decompose it first.

Reply:
The check for det(a) was aimed on catching a non-updated `BoundMatrix aMatrix = BoundMatrix::Zero();` but this is now more general and can also catch other cases.

TODO: find mathematical description of a system, that leads to a non-invertible aMatrix

Reply:
https://github.com/acts-project/acts/blob/9ba6e127b845dac1458994426c4a2920da5728f5/Core/include/Acts/Utilities/AlgebraHelpers.hpp#L188

Reply:
I now use `safeInverse(m)` for the 4x4-case.

For other cases we can not use it since it builds on `Eigen::MatrixBase< Derived >::computeInverseWithCheck`.
"This is only for fixed-size square matrices of size up to 4x4."
https://eigen.tuxfamily.org/dox-devel/classEigen_1_1MatrixBase.html#title23

I added a `@note` to `safeInverse(m)`.

Reply:
Should I maybe also create a `safeInverseLarge(m)` function in the algebra helpers using the FullPivLU? Then we could also do inverses of 5x5 and larger in the same style.

Reply:
interesting - I think we could extend it by checking the matrix size at compile time. but lets do that in another pr